### PR TITLE
Slide away UI elements on scroll

### LINKED ISSUE #1139: Slide away the search bar when scrolling down, slide it back in when scrolling up
This would be an intermediate step on the way to #523. It would already be a significant UI improvement in itself, and should be relatively easy to implement.

### PR DESCRIPTION
**Pull request by Jaifroid**
Closes #1139.

It's still a bit rough, but working in essence.

### CODE REVIEW

**Comment by Jaifroid** [comment]
Seems ready. Checked on IE11 and Firefox OS. Need to check on Edge Legacy.

**Comment by Jaifroid** [comment]
@Rishabhg71 Would you be able to test this and see if you think UX is good?

Any comments on the code would be appreciated, but optional!

If you don't have any comments and are happy with it, please approve. Thanks!

File: www/js/app.js [hunk 1]
```diff
@@ -87,6 +87,66 @@ appstate['search'] = {
...
+}
+
+let throttle = false;
+
+// Throttles the slide-away function
+function scroller () {
+    if (throttle) return;
+    // If the user has scrolled to the bottom of the article, stop processing
+    // const articleDocument = articleWindow.document.documentElement;
+    // if (articleDocument.scrollTop === (articleDocument.scrollHeight - articleDocument.clientHeight)) return;
```
Comment (Jaifroid) [inline www/js/app.js:141]: No longer necessary. Remove commented-out code.

**Comment by Jaifroid** [comment]
PS, just so you're aware, sliding away of header/footer is disabled if the search bar (prefix) has focus. Since prefix is automatically focused on loading the landing page of a ZIM, it means that the landing page won't show the scroll effect unless you click somewhere in the article. Subsequent articles will work without the extra click.

**Comment by Jaifroid** [comment]
On Edge 15 (on Windows Mobile), the header remains hidden under the iframe until the user scrolls back to the very top of the document. I probably need to set the zindex to 1.

**Comment by Rishabhg71** [comment]
@Jaifroid 
image
I have seen this behavior a lot of times usually refreshing the page the load it up properly. This only happens when running the vite server. Any idea why + Am I the only one getting this behavior?
try visiting this URL (GH codespaces) https://musical-yodel-xvpjvv4pwpvcprpp-5173.app.github.dev/www/index.html

**Comment by Jaifroid** [comment]
Yes, that's a known issue with the vite server, and is usually fixed by ensuring Bypass AppCache setting is turned ON. When it's on, and when using vite server, it's not necessary to refresh the page when you make a change, because the server forces the browser to refresh. However, if you do get the disordered display, then a refresh will fix it, or sometimes Ctrl-Shift-R with devtools in focus.

This is nothing to do with this PR/branch, it's because Vite uses [esbuild](https://esbuild.github.io/) as a fast bundler when using Hot Module Replacement, whereas it uses rollup for production builds (but I don't actually use Vite's version of rollup, because it's configured way too aggressively -- we just use pure rollup for our bundling).

**Comment by Jaifroid** [comment]
You can use http-server instead if it's bothersome (`npm run web-server`).

**Comment by Rishabhg71** [comment]
> This is nothing to do with this PR/branch, it's because Vite uses [esbuild](https://esbuild.github.io/) as a fast bundler when using Hot Module Replacement, whereas it uses rollup for production builds (but I don't actually use Vite's version of rollup, because it's configured way too aggressively -- we just use pure rollup for our bundling).

I had an idea it was a problem with vite because the normal HTTP server works now coming to the main I feel UI/UX is good but IMHO I found it annoying that every time I scroll up and down. A simple solution is to only show the header/footer if a minimum distance has been crossed by the scroll wheel.
On a side note, There is a weird bug on Brave where the header hides but leaves that red area It might be an issue on Brave or all Chromium-based browsers. I know we don't officially support Brave so maybe we can leave it

https://github.com/kiwix/kiwix-js/assets/47270636/22a2f7db-502f-47de-bc02-581c0197a625

**Comment by Jaifroid** [comment]
> On a side note, There is a weird bug on Brave where the header hides but leaves that red area It might be an issue on Brave or all Chromium-based browsers. I know we don't officially support Brave so maybe we can leave it

Oh, well it's definitely not the case with all Chromium browsers, as it doesn't show on Edge. I'll check on Chrome. However, I should try to fix this, unless it really is a bug with Brave. I'll check.

> A simple solution is to only show the header/footer if a minimum distance has been crossed by the scroll wheel.

Yes, I could try that. However, I think that should apply to scrolling down (in fact I did make it that the user needs to scroll at least 50 pixels from the top, but not on subsequent scrolls). On scrolling up, it's more likely the user wants to see the header so they can go somewhere else or search for something, so they might be annoyed if it doesn't appear straight away.

Anyway, thanks for checking, and I'll see if I can make improvements.

(By the way, if a user is really annoyed, they can always turn the feature off in Configuration.)

**Comment by Jaifroid** [comment]
Quick question, do you think this feature should be off by default, and users have to turn it on if they want it?

I was kind of taking my cue from the Android app where it's always on (and I don't think it can be turned off).

**Comment by Rishabhg71** [comment]
> Quick question, do you think this feature should be off by default, and users have to turn it on if they want it?

Turning this feature on by default makes more sense. Most of the users don't care for these minor inconveniences and also get used to them after some time. One more suggestion you can try playing around with the animation time I feel it's a little on the slow side

**Comment by Jaifroid** [comment]
So I've done these changes:

* Animation reduced to 0.3s instead of 0.5s. You're right, this makes it feel more responsive.
* User has to scroll down at least 50 pixels (from any starting position) to activate hiding. This means that really slow scrolling won't activate it (due to my intentional throttling of the onscroll event), but I think that's quite good because the result is that you need a slightly more decisive scroll in order to activate it. Keyboard down-arrow mostly activates it, but occasionally only after second down-arrow. Scrolling up doesn't have the 50 pixel margin, because users will usually be doing that after they've finished reading and want to show the header/footer so they can navigate.
* I now only apply the z-index and position: relative workaround in IE11 and Edge Legacy (only Edge Legacy actually needs it).
* I think the last change (z-index) has fixed the Brave bug. At least, I can't reproduce it now using portable Brave (https://portapps.io/app/brave-portable/) installed in Windows Sandbox. I believe having changed the position of the header from static to relative is what caused the Brave bug. It's back to static for all browsers except IE11 and Edge Legacy.

File: www/js/init.js [hunk 2]
```diff
@@ -82,6 +83,8 @@ params['storeType'] = getBestAvailableStorageAPI();
 // The key prefix used by the settingsStore.js (see comment there for explanation), but we also need it below
 params['keyPrefix'] = 'kiwixjs-';
 params['hideActiveContentWarning'] = getSetting('hideActiveContentWarning') === true;
+// A parameter to determine whether to slide away the header and footer when scrolling (feaults to true)
```
Comment (Jaifroid) [inline www/js/init.js:86]: Typo -> defaults

**Comment by Jaifroid** [comment]
@Rishabhg71 If you're happy that the changes address the issues you identified, I'd be grateful if you could approve the PR! Thanks for helping to improve this PR -- it's always good to have a second opinion, especially for UI changes which can be quite contentious for users.

**Review by Rishabhg71 — APPROVED** [review]

**Comment by Jaifroid** [comment]
Last tests:

- [ ] Firefox 64 JQuery and SW modes
- [ ] FFOS double-check
- [ ] IE11 double-check
- [ ] As a Chromium extension
- [ ] As a Firefox extension

**Comment by Jaifroid** [comment]
Encountering a bug on Firefox OS whereby links clicked with a mouse stop working after a few clicks. Touch still works. Very odd. Need to check if this is also happening on main, to see if this PR caused it.

**Comment by Jaifroid** [comment]
FFOS bug appears to be triggered when the transition is set on the iframe. It appears to cause some href elements in the document to lose their click event listeners, though touch still works.. It doesn't happen if the feature is turned off.

It might be time to implement the first bullet point of #725, which is an overdue simplification of link handling, and probably really quick to do. I'll test.

**Comment by Jaifroid** [comment]
OK, so I've determined that #732 doesn't fix this. It's caused by a weird bug whereby the `event.target` of a clicked link no longer points to the right bit of HTML due to the iframe having been shifted up by several pixels (by the transform translateY style having been added and shifting the iframe up). This of course should have nothing to do with the document inside the iframe, so it's definitely a framework bug.

I tried restoring the position of the iframe immediatley on click, but this is too late, and doesn't change what the event.target points to.

Now there are two mitigating factors:

1. Real FFOS devices are touchscreen devices, and at least in the simulator, touch seems to provide the correct information, and is unaffected;
2. It can be turned off.

So, I propose to add a check for FFOS, and NOT turn this feature on by default for that platform. When user turns it on anyway, a warning can be added if it's FFOS, to use at own risk.

**Pull request merged**